src/Native/Unix/System.Native/pal_networking.c

```diff
@@ -1268,7 +1268,12 @@ int32_t SystemNative_SendMessage(intptr_t socket, MessageHeader* messageHeader,
     ConvertMessageHeaderToMsghdr(&header, messageHeader, fd);
 
     ssize_t res;
+#if defined(__APPLE__) && __APPLE__
+    // possible OSX kernel bug:  #31927
+    while ((res = sendmsg(fd, &header, socketFlags)) < 0 && (errno == EINTR || errno == EPROTOTYPE));
```

Comment:
Is it possible this could end up in an infinite loop? EPROTOTYPE isn't documented, so what's to say that if it's returned once from a call it won't be returned from the next call and every subsequent call?

Reply:
Since this is race condition it should not be permanent. This mimic https://github.com/libuv/libuv/commit/406961385d3560aa6ceba771e36c0ed11b6a57a0

It seems like we are hitting this in Kestrel test runs so we should be able to 1. validate fix and 2. any infinite loops should be quite visible. 
@halter73  may provide more insights.

Reply:
I'm definitely not an expert on this, but the best resource I found related to this issue was from [this blog post](http://erickt.github.io/blog/2014/11/19/adventures-in-debugging-a-potential-osx-kernel-bug/) by @erickt which I found referenced in the description of https://github.com/gevent/gevent/pull/1033.

Here's the conclusion to save a click:

> I believe we find that behavior in the actual TCP syscall file, bsd/netinet/tcp_usrreq.c:

```C
static int
tcp_usr_send(struct socket *so, int flags, struct mbuf *m,
     struct sockaddr *nam, struct mbuf *control, struct proc *p)
{
	int error = 0;
	struct inpcb *inp = sotoinpcb(so);
	struct tcpcb *tp;
	uint32_t msgpri = MSG_PRI_DEFAULT;
#if INET6
	int isipv6;
#endif
	TCPDEBUG0;

	if (inp == NULL || inp->inp_state == INPCB_STATE_DEAD
#if NECP
		|| (necp_socket_should_use_flow_divert(inp))
#endif /* NECP */
		) {
		/*
		 * OOPS! we lost a race, the TCP session got reset after
		 * we checked SS_CANTSENDMORE, eg: while doing uiomove or a
		 * network interrupt in the non-splnet() section of sosend().
		 */
		if (m != NULL)
			m_freem(m);
		if (control != NULL) {
			m_freem(control);
			control = NULL;
		}

		if (inp == NULL)
			error = ECONNRESET;	/* XXX EPIPE? */
		else
			error = EPROTOTYPE;
		tp = NULL;
		TCPDEBUG1();
		goto out;
	}
...
```

> I believe that comment explains everything we’re seeing. If we trigger a send while the kernel is in the middle of tearing down the socket, it returns EPROTOTYPE. This then looks to be an error we could retry. Once the socket is fully torn down, it should eventually return the proper EPIPE.

tl:dr; I don't think this could end up in an infinite loop since the race occurs during a presumably-transient socket teardown, but don't take my word for it

Reply:
@halter73 
> but don't take my word for it

:) Under some conditions, this can result in an infinite loop.  Specifically, there's the

```|| (necp_socket_should_use_flow_divert(inp))```

block, which I guess points towards a specific kind of network-extension called a TransparentProxy extension (flow divert rings similarity to divert sockets).  If such extension is enabled, the EPROTOTYPE error will not be transient when a write() call is issued to a socket that was in a connected state before the extension was enabled.

See https://developer.apple.com/forums/thread/681135 for some info.  There's also a fix to openssh referenced in the above post that guards against write errors.  I am hitting this issue in Signal.app that uses Electron -> libuv, and the issue exists in libuv as well.  It would be prudent to guard against the potential infinite loop.

Reply:
@jvimal-eg thanks for your comment but note that since this PR occurred this repo became only for servicing. You may want to open an issue or discussion in dotnet/runtime.

Reply:
Thanks @danmoseley. I created https://github.com/dotnet/runtime/issues/63291.

Reply:
appreciate it @jvimal-eg